File: backend/src/Playbook/Adapters/LoaderMcpExecutor.php
<?php
declare(strict_types=1);
namespace Quantis\AIPortfolioAssistant\Playbook\Adapters;

use Quantis\AIPortfolioAssistant\Playbook\McpExecutorInterface;
use Quantis\AIPortfolioAssistant\Services\MCPToolsLoader;

/**
 * Adapts MCPToolsLoader (DB-backed MCP tool cache + JSON-RPC proxy) to the
 * playbook interpreter's McpExecutorInterface.
 *
 * The loader keys its tools by `mcp_<tool_name>` and stores per-tool the
 * server it came from (`server_name`). This adapter re-keys that map as
 * "<server_slug>.<tool_name>" (e.g. "okta.search_users") so playbooks can
 * address tools the same way regardless of which MCP server hosts them.
 *
 * The caller is responsible for having already called
 * `$loader->loadToolsForUser(...)` before constructing this adapter.
 */
final class LoaderMcpExecutor implements McpExecutorInterface
{
    public function __construct(private readonly MCPToolsLoader $loader)
    {
    }

    public function call(string $server, string $tool, array $args): array
    {
        $entry = $this->findEntry($server, $tool);
        if ($entry === null) {
            return ['ok' => false, 'error' => "MCP tool '{$server}.{$tool}' not found"];
        }

        $result = $this->loader->executeTool('mcp_' . $entry['original_name'], $args);

        if (isset($result['error']) && $result['error'] === true) {
            return ['ok' => false, 'error' => $result['message'] ?? 'MCP tool execution failed'];
        }

        return ['ok' => true, 'result' => $result];
    }

    public function availableTools(): array
    {
        $tools = [];
        foreach ($this->loader->getTools() as $entry) {
            $key = $this->slug($entry['server_name']) . '.' . $entry['original_name'];
            $tools[$key] = [
                'description' => $entry['description'] ?? '',
                'input_schema' => $this->decodeSchema($entry['input_schema_json'] ?? null),
            ];
        }
        return $tools;
    }

    /**
     * Find the loader's tool entry matching "<slug(server_name)>.<original_name>".
     */
    private function findEntry(string $server, string $tool): ?array
    {
        foreach ($this->loader->getTools() as $entry) {
            if ($this->slug($entry['server_name']) === $server && $entry['original_name'] === $tool) {
                return $entry;
            }
        }
        return null;
    }

    private function slug(string $serverName): string
    {
        return strtolower((string) preg_replace('/[^a-z0-9]+/i', '_', $serverName));
    }

    private function decodeSchema(?string $json): array
    {
        if ($json === null || $json === '') {
            return ['type' => 'object', 'properties' => []];
        }
        $decoded = json_decode($json, true);
        return is_array($decoded) ? $decoded : ['type' => 'object', 'properties' => []];
    }
}

File: backend_python/app/playbook/adapters/loader_mcp_executor.py
"""Port of backend/src/Playbook/Adapters/LoaderMcpExecutor.php (81 lines).

Adapts MCPToolsLoader (DB-backed MCP tool cache + JSON-RPC proxy) to the
playbook interpreter's McpExecutorInterface.

The loader keys its tools by `mcp_<tool_name>` and stores per-tool the
server it came from (`server_name`). This adapter re-keys that map as
"<server_slug>.<tool_name>" (e.g. "okta.search_users") so playbooks can
address tools the same way regardless of which MCP server hosts them.

The caller is responsible for having already called
`loader.loadToolsForUser(...)` before constructing this adapter.
"""
from __future__ import annotations

import json
import re
from typing import Optional

from app.playbook.mcp_executor_interface import McpExecutorInterface

_SLUG_RE = re.compile(r'[^a-z0-9]+', re.IGNORECASE)


class LoaderMcpExecutor(McpExecutorInterface):
    def __init__(self, loader):
        self.loader = loader

    def call(self, server: str, tool: str, args: dict) -> dict:
        entry = self._findEntry(server, tool)
        if entry is None:
            return {'ok': False, 'error': f"MCP tool '{server}.{tool}' not found"}

        result = self.loader.executeTool('mcp_' + entry['original_name'], args)

        if isinstance(result, dict) and result.get('error') is True:
            return {'ok': False, 'error': result.get('message') if result.get('message') is not None else 'MCP tool execution failed'}

        return {'ok': True, 'result': result}

    def availableTools(self) -> dict:
        tools = {}
        for entry in self.loader.getTools().values():
            key = self._slug(entry['server_name']) + '.' + entry['original_name']
            tools[key] = {
                'description': entry.get('description') if entry.get('description') is not None else '',
                'input_schema': self._decodeSchema(entry.get('input_schema_json')),
            }
        return tools

    def _findEntry(self, server: str, tool: str) -> Optional[dict]:
        """Find the loader's tool entry matching "<slug(server_name)>.<original_name>"."""
        for entry in self.loader.getTools().values():
            if self._slug(entry['server_name']) == server and entry['original_name'] == tool:
                return entry
        return None

    def _slug(self, serverName: str) -> str:
        return _SLUG_RE.sub('_', serverName).lower()

    def _decodeSchema(self, json_str: Optional[str]) -> dict:
        if json_str is None or json_str == '':
            return {'type': 'object', 'properties': {}}
        try:
            decoded = json.loads(json_str)
        except (ValueError, TypeError):
            return {'type': 'object', 'properties': {}}
        return decoded if isinstance(decoded, (dict, list)) else {'type': 'object', 'properties': {}}
